### PR TITLE
#1 Init resources script

### PR DESCRIPTION
**Pull request by lojzatran**
Fixes: https://github.com/commercetools/commercetools-adyen-integration/issues/1

This PR adds a script for creating custom types for [payment](https://github.com/commercetools/commercetools-adyen-integration/pull/4/files#diff-95f17e4d1b4a4ceeb9153148013399ba) and [interface interaction](https://github.com/commercetools/commercetools-adyen-integration/pull/4/files#diff-ff8cc44e0820821822faf994a2ae0eb6) and [API extensions](https://github.com/commercetools/commercetools-adyen-integration/pull/4/files#diff-2f64d0d89975bcdea0ecad259ec955c3). There's a [script for creating all above resources at once](https://github.com/commercetools/commercetools-adyen-integration/pull/4/files#diff-17f523f051330c32c5ddd8d648b36d43).

The server and controller parts are not completed yet, it's here only for testing now. I will create another PRs where they will be improved.

Additionally, there are some config files in order to start the project.

### CODE REVIEW

File: extension/src/config/config.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  }
+}
+
+function getCTPEnvCredentials () {
+  return {
+    projectKey: process.env.CTP_PROJECT_KEY,
+    clientId: process.env.CTP_CLIENT_ID,
+    clientSecret: process.env.CTP_CLIENT_SECRET,
+    apiUrl: process.env.CTP_API_URL,
+    authUrl: process.env.CTP_AUTH_URL
```
Comment (LEQADA) [inline extension/src/config/config.js]: I think we can hardcode API and Auth URLs. They will never change

File: .gitignore [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+# Compiled binary addons (https://nodejs.org/api/addons.html)
+build/Release
+
+# Dependency directories
+node_modules/
+jspm_packages/
+
+# TypeScript v1 declaration files
+typings/
```
[the remark this answers is not in the document]
Comment (lojzatran) [reply .gitignore]: Autogenerated file.

File: extension/src/config/config.js (hunk 1 above)
Comment (lojzatran) [reply extension/src/config/config.js]: https://github.com/commercetools/commercetools-adyen-integration/pull/5/commits/bf2f7227796b91c52b6e104ebb7c5fa8d31ff68d

File: extension/scripts/initApiExtensions.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+const apiExtensionDrafts = require('../resources/api-extensions.json')
+
+async function initApiExtensions (ctpClient, ctpAdyenIntegrationBaseUrl) {
+  await Promise.map(apiExtensionDrafts.values(), async (uncompiledDraft) => {
+    try {
+      const extension = _.template(JSON.stringify(uncompiledDraft))({ ctpAdyenIntegrationBaseUrl })
+      await ctpClient.create(ctpClient.builder.extensions, extension)
+    } catch (e) {
+      console.log('Error when creating interface interaction custom type, skipping...', JSON.stringify(e))
```
Comment (LEQADA) [inline extension/scripts/initApiExtensions.js]:
```suggestion
      console.log('Error when creating API extension, skipping...', JSON.stringify(e))
```

File: extension/resources/api-extensions.json [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+      {
+        "resourceTypeId": "payment",
+        "actions": [
+          "Create",
+          "Update"
+        ]
+      }
+    ]
+  }
+]
```
Comment (LEQADA) [inline extension/resources/api-extensions.json]: It looks like we will have only 1 extension. So maybe let's make it an object?

File: extension/scripts/initApiExtensions.js (hunk 3 above)
Comment (LEQADA) [reply extension/scripts/initApiExtensions.js]: I think it should be logged as an error. But here you don't have a logger. So I'd write it as a `console.error()` for now.

File: extension/src/api/payment/payment.controller.js [hunk 5]
```diff
@@ -0,0 +1,17 @@
+const httpUtils = require('../../utils')
+
+async function handlePayment (request, response) {
+  const paymentObject = await _getPaymentObject(request)
+  //const handler = this._getPaymentHandler(paymentObject)
+  //handler.execute(paymentObject)
```
Comment (LEQADA) [inline extension/src/api/payment/payment.controller.js]: do you need these?

File: extension/src/api/payment/payment.controller.js [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+  return httpUtils.sendResponse(response)
+}
+
+async function _getPaymentObject (request) {
+  return httpUtils.collectRequestData(request)
+}
+
+function _getPaymentHandler (paymentObject) {
+}
```
Comment (LEQADA) [inline extension/src/api/payment/payment.controller.js]: empty function?

File: extension/src/config/config.js (hunk 1 above)
Comment (LEQADA) [reply extension/src/config/config.js]: No sure we need it global either. I think you can hardcode it on the client creation step

File: extension/src/config/config.js [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+  }
+}
+
+function getFileConfig () {
+  let fileConfig = {}
+  try {
+    fileConfig = require(configPath) // eslint-disable-line
+  } catch (e) {
+    // config file was not provided
```
Comment (LEQADA) [inline extension/src/config/config.js]: log it?

File: extension/src/init.js [hunk 8]
```diff
@@ -0,0 +1,8 @@
+const init = require('./server.js')
+require('./config/config')
+
+const port = 8080
```
Comment (LEQADA) [inline extension/src/init.js]: It can be an env variable

File: index.js [hunk 9]
```diff
@@ -0,0 +1,8 @@
+const http = require('http')
+
+http.createServer(function (request, response) {
```
Comment (LEQADA) [inline index.js]: you can make it an arrow function as you did everywhere

File: package.json [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+  "dependencies": {
+    "bluebird": "3.5.3",
+    "@commercetools/api-request-builder": "5.0.0",
+    "@commercetools/sdk-client": "2.0.0",
+    "@commercetools/sdk-middleware-auth": "5.0.0",
+    "@commercetools/sdk-middleware-http": "5.0.0",
+    "@commercetools/sdk-middleware-queue": "2.0.0",
+    "@commercetools/sync-actions": "3.0.0",
+    "lodash": "4.17.11",
+    "node-fetch": "2.3.0"
```
Comment (LEQADA) [inline package.json]: why all versions are fixed? Is there a specific idea behind it?

File: extension/src/api/payment/payment.controller.js (hunk 5 above)
Comment (lojzatran) [reply extension/src/api/payment/payment.controller.js]: No, I don't.

File: extension/src/api/payment/payment.controller.js (hunk 6 above)
Comment (lojzatran) [reply extension/src/api/payment/payment.controller.js]: Yes, but it will be filled in the next PR.

File: index.js (hunk 9 above)
Comment (lojzatran) [reply index.js]: This file is removed in the next PRs, so I will not change it here.

File: package.json (hunk 10 above)
Comment (lojzatran) [reply package.json]: No reason. Are there reasons not to use fix versions?

File: extension/src/config/config.js (hunk 1 above)
Comment (lojzatran) [reply extension/src/config/config.js]: https://github.com/commercetools/commercetools-adyen-integration/pull/5/commits/a5f5edc41e3a31d7fcdb78b20515a5b131b9446a#diff-bb1f7f72b7640d409011bd3d2371b1b7

File: extension/scripts/initApiExtensions.js (hunk 3 above)
Comment (lojzatran) [reply extension/scripts/initApiExtensions.js]: https://github.com/commercetools/commercetools-adyen-integration/pull/4/commits/973138fc68d46f5c3c1cdd735c0f7e8932d55bc2

File: extension/resources/api-extensions.json (hunk 4 above)
Comment (lojzatran) [reply extension/resources/api-extensions.json]: https://github.com/commercetools/commercetools-adyen-integration/pull/4/commits/77f4e9d77bc011cd79fc1f0f0b9d70f4babaeab6

File: extension/src/config/config.js (hunk 7 above)
Comment (lojzatran) [reply extension/src/config/config.js]: This method is removed: https://github.com/commercetools/commercetools-adyen-integration/pull/4/commits/0f21fef327bad8c18c558665e226e79dd5f01816

File: extension/src/init.js (hunk 8 above)
Comment (lojzatran) [reply extension/src/init.js]: https://github.com/commercetools/commercetools-adyen-integration/pull/4/commits/71a1028fa7b6a8ac7f47f45c09c18b2d92a6dc53

File: .gitignore (hunk 2 above)
[the remark this answers is not in the document]
Comment (LEQADA) [reply .gitignore]: don't you want to clean it up?

[the remark this answers is not in the document]
Comment (lojzatran) [reply .gitignore]: No?

File: package.json (hunk 10 above)
Comment (LEQADA) [reply package.json]: To get all versions you can use `npm ls --depth=0`.
`^` is useful when you don't want to update all dependencies one by one

Comment (lojzatran) [reply package.json]: https://github.com/commercetools/commercetools-adyen-integration/pull/4/commits/536971d0e3373203fed242299f5fb23e34407de7

**Pull request merged**